Deberta v2 support

vrdn-23 opened the pull request:
@kozistr thanks again for agreeing to help out!

vrdn-23 committed:
Rename config and remove dropout

vrdn-23 committed:
Remove NER and id2label

vrdn-23 committed:
Load debertav2 classification head but not implement trait/forward

vrdn-23 committed:
[WIP] loaded weights but failing during warmup

vrdn-23 committed:
IT WORKS!! but not for batches :(

kozistr commented on backends/candle/src/models/debertav2.rs:
```diff
@@ -0,0 +1,1490 @@
+    Gelu,
+    GeluApproximate,
+    Relu,
+}
```
We could use [HiddenAct](https://github.com/huggingface/text-embeddings-inference/blob/6e900afba71821fdf250e380d7da1f5a6e5e7e27/backends/candle/src/layers/linear.rs#L7) enum instead of creating a new one!

vrdn-23 replied:
I added that a custom serializer but unsure if there is a better way to do this than having 2 different functions when the Pooler_Act is optional v/s not.

Any ideas?

kozistr commented on backends/candle/src/models/debertav2.rs:
```diff
@@ -0,0 +1,1490 @@
+    pub intermediate_size: usize,
+    pub hidden_act: HiddenAct,
+    pub hidden_dropout_prob: f64,
+    pub attention_probs_dropout_prob: f64,
```
We mostly don't use a dropout layer for the inference stage, so maybe we could remove these configs

kozistr commented on backends/candle/src/models/debertav2.rs:
```diff
@@ -0,0 +1,1490 @@
+pub struct DebertaV2DisentangledSelfAttention {
+    config: DebertaV2Config,
+    num_attention_heads: usize,
+    query_proj: candle_nn::Linear,
```
`Linear` is frequently used over the codes, so I guess it'd be better to use `Linear`, rather than `candle_nn::Linear`.

kozistr commented on backends/candle/src/models/debertav2.rs:
```diff
@@ -0,0 +1,1490 @@
+
+// https://github.com/huggingface/transformers/blob/78b2929c0554b79e0489b451ce4ece14d265ead2/src/transformers/models/deberta_v2/modeling_deberta_v2.py#L605
+pub struct DebertaV2DisentangledSelfAttention {
+    config: DebertaV2Config,
```
it looks like `config` is not used, i guess including only the necessary configs would be great in terms of compactness.

kozistr commented on backends/candle/src/models/debertav2.rs:
```diff
@@ -0,0 +1,1490 @@
+
+        let mut scale_factor: usize = 1;
+
+        if self.config.pos_att_type.iter().any(|s| s == "c2p") {
```
how about creating a variable like `is_c2p_attn` to predefine the pos_att_type instead of searching the attn type on the fly?

or we could do this in the deserializer

kozistr commented on backends/candle/src/models/debertav2.rs:
```diff
@@ -0,0 +1,1490 @@
+            }
+        }
+    }
+}
```
EOF

kozistr commented on backends/candle/src/models/debertav2.rs:
```diff
@@ -0,0 +1,1490 @@
+    embed_proj: Option<candle_nn::Linear>,
+}
+
+impl DebertaV2Embeddings {
```
looks like, in TEI, each `impl` has a `span`.

kozistr commented on backends/candle/src/models/debertav2.rs:
```diff
@@ -0,0 +1,1490 @@
+
+    pub fn forward(
+        &self,
+        input_ids: Option<&Tensor>,
+        token_type_ids: Option<&Tensor>,
+        position_ids: Option<&Tensor>,
```
We always have their values (`input_ids`, `token-type_ids`, `position_ids`), so I think `Tensor` type would be enough

kozistr commented on backends/candle/src/models/debertav2.rs:
```diff
@@ -0,0 +1,1490 @@
+// https://github.com/huggingface/transformers/blob/78b2929c0554b79e0489b451ce4ece14d265ead2/src/transformers/models/deberta_v2/modeling_deberta_v2.py#L72
+struct XSoftmax {}
+
+impl XSoftmax {
```
I guess, we don't need this layer! Just add the attention mask to the input like [this](https://github.com/huggingface/text-embeddings-inference/blob/6e900afba71821fdf250e380d7da1f5a6e5e7e27/backends/candle/src/models/modernbert.rs#L261)

kozistr commented on backends/candle/src/models/debertav2.rs:
```diff
@@ -0,0 +1,1490 @@
+            None,
+        )?;
+
+        let encoder_attention_mask = attention_mask.as_ref().cloned().unwrap_or_else(|| Tensor::ones(shape, DType::I64, &self.device).unwrap());
```
How about creating the attention mask when batch_size is 1, at line 1202, if we eventually need one?

And it seems like the dtype of the attention mask is float32 when batched input, while int64 in the single input case.

kozistr reviewed (commented):
great work! I've just left some reviews, and will get back to you when I have some time on my hands :)

vrdn-23 committed:
fix merge conflicts

vrdn-23 committed:
interim review changes

vrdn-23 commented:
He @kozistr, I finally got some time to go through this and started addressing your comments. I will mostly raise a PR against the main repo this week, so hoping to get more feedback from you there!

vrdn-23 committed:
interim changes again

vrdn-23 committed:
fix linting

vrdn-23 committed:
refactor our xsoftmax

vrdn-23 committed:
linting

vrdn-23 committed:
Merge remote-tracking branch 'origin/main' into vidamoda/deberta-v2

vrdn-23 commented:
I've addressed all the comments in this PR. closing this in favor of https://github.com/huggingface/text-embeddings-inference/pull/746
